consensus/state.go

```diff
@@ -768,7 +775,26 @@ func (cs *State) receiveRoutine(maxSteps int) {
 
 			// handles proposals, block parts, votes
 			// may generate internal events (votes, complete proposals, 2/3 majorities)
-			cs.handleMsg(mi)
+			err := cs.handleMsg(mi)
+			if err != nil {
+				// if error is due to misbehaviour, disconect from peer
+				// TODO: is there a cleaner way to do these checks?
+				if errors.Is(err, ErrInvalidProposalSignature) ||
+					errors.Is(err, types.ErrVoteInvalidSignature) ||
+					errors.Is(err, types.ErrVoteInvalidValidatorAddress) ||
+					errors.Is(err, types.ErrVoteInvalidValidatorIndex) {
+
+					// tell reactor to disconnect from peer
+					cs.peerInfoQueue <- badPeerInfo{
+						Error:  err,
+						PeerID: mi.PeerID,
+					}
+				} else {
+					// other errors don't necessarily mean the peer is bad,
+					// so there is nothing to do
+					// see https://github.com/tendermint/tendermint/issues/2871
+				}
+			}
```

Comment:
handleMsg is a method defined on the State type. It returns an error which, apparently, in certain cases, causes the caller to send something on a Go channel defined as a field on that same State type. Is there any reason that this work is not done by the handleMsg method directly? Maybe there is a reason! and so maybe this should in fact be done by the caller like this here code suggests. But then you want like

```go
if err := cs.wal.Write(mi); err != nil {
	cs.Logger.Error("failed writing to WAL", "err", err)
	// IS IT VALID TO CARRY ON FROM THIS POINT?? I DOUBT IT?? 
	// PROBABLY NEED TO CONTINUE HERE?? I MEAN DANG DUDE
}

var (
	handleErr                        = cs.handleMsg(mi)
	isErrInvalidProposalSignature    = errors.Is(handleErr, ErrInvalidProposalSignature)
	isErrVoteInvalidSignature        = errors.Is(handleErr, types.ErrVoteInvalidSignature)
	isErrVoteInvalidValidatorAddress = errors.Is(handleErr, types.ErrVoteInvalidValidatorAddress)
	isErrVoteInvalidValidatorIndex   = errors.Is(handleErr, types.ErrVoteInvalidValidatorIndex)
	isMisbehavior                    = isErrInvalidProposalSignature || isErrVoteInvalidSignature || isErrVoteInvalidValidatorAddress || isErrVoteInvalidValidatorIndex
)
switch {
case isMisbehavior:
	cs.peerInfoQueue <- badPeerInfo{Error: handleErr, PeerID: mi.PeerID}
default:
	// PROBABLY AT LEAST LOG SOMETHING, I MEAN LIKE WYD FAM
}
```